c/rc_lock.c

```diff
@@ -0,0 +1,462 @@
+    lock_bytes_seg->size = lock_length;
+  }
+  return CKB_SUCCESS;
+}
+
+int parse_args(ArgsType *args) {
+  int err = 0;
+  uint8_t script[SCRIPT_SIZE];
+  uint64_t len = SCRIPT_SIZE;
+  err = ckb_checked_load_script(script, &len, 0);
+  CHECK(err);
+
+  mol_seg_t script_seg;
+  script_seg.ptr = script;
+  script_seg.size = len;
+
+  mol_errno mol_err = MolReader_Script_verify(&script_seg, false);
+  CHECK2(mol_err == MOL_OK, ERROR_ENCODING);
+
+  mol_seg_t args_seg = MolReader_Script_get_args(&script_seg);
+  mol_seg_t args_bytes_seg = MolReader_Bytes_raw_bytes(&args_seg);
+  CHECK2(args_bytes_seg.size >= 1, ERROR_ARGUMENTS_LEN);
+  uint8_t flags = args_bytes_seg.ptr[0];
+  CHECK2(flags == IdentityFlagsPubkeyHash || flags == IdentityFlagsOwnerLock ||
+             flags == IdentityFlagsPubkeyHashRc ||
+             flags == IdentityFlagsOwnerLockRc,
+         ERROR_INVALID_MOL_FORMAT);
+
+  CHECK2(args_bytes_seg.size >= (1 + BLAKE160_SIZE + BLAKE2B_BLOCK_SIZE),
+         ERROR_ARGUMENTS_LEN);
+  args->id.flags = args_bytes_seg.ptr[0];
+  memcpy(args->id.blake160, args_bytes_seg.ptr + 1, BLAKE160_SIZE);
+  memcpy(args->rc_root, args_bytes_seg.ptr + 1 + BLAKE160_SIZE,
+         sizeof(args->rc_root));
+
+exit:
+  return err;
+}
+
+int load_and_hash_witness(blake2b_state *ctx, size_t start, size_t index,
+                          size_t source, bool hash_length) {
+  uint8_t temp[ONE_BATCH_SIZE];
+  uint64_t len = ONE_BATCH_SIZE;
+  int ret = ckb_load_witness(temp, &len, start, index, source);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+  if (hash_length) {
+    blake2b_update(ctx, (char *)&len, sizeof(uint64_t));
+  }
+  uint64_t offset = (len > ONE_BATCH_SIZE) ? ONE_BATCH_SIZE : len;
+  blake2b_update(ctx, temp, offset);
+  while (offset < len) {
+    uint64_t current_len = ONE_BATCH_SIZE;
+    ret = ckb_load_witness(temp, &current_len, start + offset, index, source);
+    if (ret != CKB_SUCCESS) {
+      return ret;
+    }
+    uint64_t current_read =
+        (current_len > ONE_BATCH_SIZE) ? ONE_BATCH_SIZE : current_len;
+    blake2b_update(ctx, temp, current_read);
+    offset += current_read;
+  }
+  return CKB_SUCCESS;
+}
+
+int verify_secp256k1_blake160_sighash_all(uint8_t *pubkey_hash,
+                                          uint8_t *signature_bytes) {
+  int ret;
+  uint64_t len = 0;
+  unsigned char temp[MAX_WITNESS_SIZE];
+  uint64_t read_len = MAX_WITNESS_SIZE;
+  uint64_t witness_len = MAX_WITNESS_SIZE;
+
+  /* Load witness of first input */
+  ret = ckb_load_witness(temp, &read_len, 0, 0, CKB_SOURCE_GROUP_INPUT);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
+  }
+  witness_len = read_len;
+  if (read_len > MAX_WITNESS_SIZE) {
+    read_len = MAX_WITNESS_SIZE;
+  }
+
+  /* load signature */
+  mol_seg_t lock_bytes_seg;
+  ret = extract_witness_lock(temp, read_len, &lock_bytes_seg);
+  if (ret != 0) {
+    return ERROR_ENCODING;
+  }
+  if (lock_bytes_seg.size < SIGNATURE_SIZE) {
```

Comment:
What if `lock_bytes_seg` has more bytes than `SIGNATURE_SIZE`?

Reply:
I'm not sure it's allowed or not to to append extra bytes at the end of signature.

Reply:
Personally I would opt for the strict side, and disallow extra bytes.

Reply:
The lock part (lock_bytes_seg) in witness can contain signature or proofs, or both, or none. The real size is unknown at this time.
This checking is to make sure it's long enough to hold a signature. Actually the strict checking is done at here:
```
   uint32_t read_len =
        mol2_read_at(&signature_cursor, signature_bytes, SIGNATURE_SIZE);
```